Add Vale to the Lint (Docs) workflow

ptgott opened the pull request:
Install and run Vale in the Lint (Docs) workflow in order to ensure that the documentation has a consistent style. This approach downloads a release from the Vale GitHub repository, extracts a precompiled binary, and runs it on all docs files. It assumes that all Vale warnings and errors are already resolved, so that any new ones would have been introduced by the target pull request.

This is an alternative to using reviewdog, which was behind a security incident (see #53162 for context). While reviewdog can inspect the diff introduced by a PR and limit its checks to changed lines, this change does not require reliance on an insecure dependency.

ptgott committed:
Add Vale to the Lint (Docs) workflow

Install and run Vale in the Lint (Docs) workflow in order to ensure that
the documentation has a consistent style. This approach downloads a
release from the Vale GitHub repository, extracts a precompiled binary,
and runs it on all docs files. It assumes that all Vale warnings and
errors are already resolved, so that any new ones would have been
introduced by the target pull request.

This is an alternative to using reviewdog, which was behind a security
incident (see #53162 for context). While reviewdog can inspect the diff
introduced by a PR and limit its checks to changed lines, this change
does not require reliance on an insecure dependency.

camscale commented on .github/workflows/doc-tests.yaml:
```diff
@@ -142,6 +142,26 @@ jobs:
         working-directory: 'docs'
         run: yarn markdown-lint
 
+      - name: Check out vale
```
Do you actually need the vale checkout for anything? Or is it just for the `gh release download` command? If the latter, you can run `gh release download -R errata-ai/vale ...`

camscale commented on .github/workflows/doc-tests.yaml:
```diff
@@ -142,6 +142,26 @@ jobs:
+          VALE_VERSION: 3.12.0
+        run: |
+          gh release download v${VALE_VERSION} --pattern="*Linux_64-bit*"
+          tar -xvf vale_${VALE_VERSION}_Linux_64-bit.tar.gz
```
This needs to do a check of the hash so the binaries cannot be changed maliciously. I would also download the specific file in case the wildcard ends up matching more than 1 file later for some reason.

```
VALE_FILENAME="vale_${VALE_VERSION}_Linuc_64-bit.tar.gz"
gh release download -R errata-ai/vale "v${VALE_VERSION}" --pattern="${VALE_FILENAME}"
sha256 --check <<< "3f4ea05cd1f2291b660ecf11a77cbb6b544b0f3b780604ad183f63285611321b  ${VALE_FILENAME}"
tar -xvf "${VALE_FILENAME}"
```

camscale replied:
Or perhaps add the hash as an env var alongside the version so it is clear(er) they need to be updated together:

```
env:
    VALE_VERSION: 3.12.0
    VALE_SHA256SUM: 3f4ea05cd1f2291b660ecf11a77cbb6b544b0f3b780604ad183f63285611321b
run:
...
sha256sum --check <<<"${VALE_SHA256SUM}  ${VALE_FILENAME}"
```

ptgott replied:
Thanks! I've added the integrity check.

ptgott committed:
Respond to camscale feedback

- Use the `--repo` flag instead of the `checkout` GHA job to download
  a Vale release.
- Perform an integrity check on the Vale archive.

ptgott committed:
Respond to fheinecke feedback

Make sure environment variables are quoted.

ptgott commented on .github/workflows/doc-tests.yaml:
```diff
@@ -142,6 +142,24 @@ jobs:
+      - name: Check docs style
+        working-directory: 'docs/content/current'
+        run:
+          ${GITHUB_WORKSPACE}/vale --config docs/.vale.ini docs/pages
```
Thanks!

ptgott committed:
Print a note before Vale workflow runs

Indicate that it is possible to ignore Vale rules. This way, docs
authors do not need to be blocked on edge cases.
